### PR TITLE
test(sol) add tests for fields validation

### PR DESCRIPTION
**Pull request by moalidv**

### CODE REVIEW

File: packages/sol/src/verification.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+}
+export function isValidUserBirthYear(
+  userBirthYear: number
+): FieldError.TutorTooOld | FieldError.TutorTooYoung | boolean {
+  if (new Date().getFullYear() - userBirthYear > 60)
+    return FieldError.TutorTooOld;
+  if (new Date().getFullYear() - userBirthYear < 10)
+    return FieldError.TutorTooYoung;
+  return true;
+}
```
Comment (neuodev) [inline packages/sol/src/verification.ts:39]:
```suggestion
export function isValidUserBirthYear(
  birthYear: number
): FieldError.TutorTooOld | FieldError.TutorTooYoung | boolean {
  const currentYear = new Date().getFullYear();
  const minYear = currentYear - 60;
  const maxYear = currentYear - 10;
  if (birthYear < minYear) return FieldError.UserToOld;
  if (birthYear > maxYear ) return FieldError.UserTooYoung;
  return true;
}
```

File: packages/sol/tests/verification.test.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
...l skills such as time management, study techniques, and problem-solving. These skills are not only valuable for immediate academic success but also for future educational and professional pursuits. A tutor teaches students how to approach their work systematically, breaking down larger tasks into manageable steps, which leads to better productivity and less stress.Moreover, tutoring is beneficial for students of all ages. For younger learners, a tutor helps lay a strong foundation in critical subjects like reading and math. For older students, tutors offer specialized support in more advanced topics, such as science, mathematics, or languages. Tutoring can also be crucial for exam preparation, where tutors guide students through practice tests, review key concepts, and teach test-taking strategies, helping them feel more prepared and confident.In conclusion, tutors play a key role in enhancing a student's academic journey. They provide personalized, focused instruction that helps students overcome challenges, build confidence, and achieve their educational goals. Through consistent support, a tutor not only improves a student's grades but also fosters a lifelong love for learning."
```
Comment (neuodev) [inline packages/sol/tests/verification.test.ts:117]: Don't use this entire test block. Use a package like `fakerjs` to generate fake text for you.

**Review by neuodev — CHANGES_REQUESTED** [review]

File: packages/sol/src/verification.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  return emailRegex.test(email) || FieldError.InvalidEmail;
+}
+
+export function isValidPassword(
+  password: string
+):
+  | FieldError.PasswordTooShort
+  | FieldError.PasswordMissingLetters
+  | FieldError.PasswordMissingNumbers
+  | boolean {
```
[the remark this answers is not in the document]
Comment (moalidv) [reply packages/sol/src/verification.ts:15]: Done!

**Comment by moalidv** [comment]
@neuodev 
please take a look.
there is only one test remains which is for invoice receiver 
I want some decleration about how can we check if the input value is aleardy instapay or bank account or mobile wallet or not.
and thanks in advance.

File: packages/sol/src/verification.ts [hunk 4]
```diff
@@ -0,0 +1,372 @@
+import { FieldError } from "@litespace/types";
+import { sanitizeMessage } from "./chat";
+
+export function isValidEmail(email: string): FieldError.InvalidEmail | true {
+  const emailRegex = /^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$/;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:5]: Please move any constant to `sol/src/constants.ts`. Constants should be all uppercase with `_` between words. Example: `USER_EMIAIL_REGEX`.

**Comment by moalidv** [comment]
@neuodev  
I can't find the terms for validation of the bank accounts

File: packages/sol/src/verification.ts [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+
+export function isValidTutorBio(
+  tutorBio: string
+):
+  | FieldError.TutorBioInvalidHTML
+  | FieldError.TutorBioEmpty
+  | FieldError.TutorBioTooLong
+  | true {
+  const tutorBioText = getTextFromHTMLString(tutorBio);
+  console.log("text", tutorBioText);
```
Comment (mostafakamar2308) [inline packages/sol/src/verification.ts:70]: remove it as it's not needed any more

File: packages/sol/src/verification.ts [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+  }
+  if (type === IInvoice.RecivingMethod.Instapay) {
+    if (!INSTAPAY_REGEX.test(receiver)) return FieldError.InstapayIPAInvalid;
+    return true;
+  }
+  // if (type === IInvoice.RecivingMethod.Bank) {
+  //   if (bankName === Bank.Alex) {
+
+  //   }
+  // }
```
Comment (mostafakamar2308) [inline packages/sol/src/verification.ts:361]: I see that you want to test the banks that the user provided
what do you think if we did something like this:
```
if (type === IInvoice.RecivingMethod.Bank) {
    if (!bankName) return FieldError.BankNameEmpty;
    if (!Object.values(Bank).includes(bankName)) {
    return FieldError.BankNameInvalid;
    }
    return true
  }
```
Don't forget to implement `FieldError.BankNameEmpty` and `FieldError.BankNameInvalid`

File: packages/types/src/invoice.ts [hunk 7]
```diff
@@ -116,3 +116,9 @@ export type CancelApiPayload = {
 };
 
 export type FindInvoicesParams = IFilter.Pagination & { userId?: number };
+
+export enum RecivingMethod {
```
Comment (mostafakamar2308) [inline packages/types/src/invoice.ts:120]: Very Nice 🚀

**Review by mostafakamar2308 — CHANGES_REQUESTED** [review]
Thank you for your hard work ❤️ 
Can you also take a look on some of my comments?

File: packages/sol/src/verification.ts [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+  const minYear = currentYear - 60;
+  const maxYear = currentYear - 10;
+
+  if (userBirthYear < minYear) return FieldError.TutorTooOld;
+  if (userBirthYear > maxYear) return FieldError.TutorTooYoung;
+  return true;
+}
+
+export function isValidTutorBio(
+  tutorBio: string
```
Comment (neuodev) [inline packages/sol/src/verification.ts:63]:
```suggestion
  bio: string
```

File: packages/sol/src/verification.ts [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+):
+  | FieldError.TutorBioInvalidHTML
+  | FieldError.TutorBioEmpty
+  | FieldError.TutorBioTooLong
+  | true {
+  const tutorBioText = getTextFromHTMLString(tutorBio);
+  console.log("text", tutorBioText);
+  if (!tutorBioText.length) return FieldError.TutorBioEmpty;
+  if (!REGEXFORHTML.test(tutorBio)) return FieldError.TutorBioInvalidHTML;
+  if (tutorBioText.length > 60) return FieldError.TutorBioTooLong;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:73]: Always remember not to use magic numbers.

Comment (neuodev) [inline packages/sol/src/verification.ts:73]:
```suggestion
  if (!tutorBioText.length) return FieldError.EmptyBio;
  if (!REGEXFORHTML.test(tutorBio)) return FieldError.InvalidBio;
  if (tutorBioText.length > 60) return FieldError.BioToLong;
```

File: packages/sol/src/verification.ts [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+  return true;
+}
+
+export function isValidTutorAbout(
+  tutorAbout: string
```
Comment (neuodev) [inline packages/sol/src/verification.ts:78]:
```suggestion
  about: string
```

File: packages/sol/src/verification.ts (hunk 8 above)
Comment (neuodev) [inline packages/sol/src/verification.ts:63]: Tutor's bio is a plain text. It is not a raw HTML.

File: packages/sol/src/verification.ts [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+):
+  | FieldError.TutorAboutInvalidHTML
+  | FieldError.TutorAboutEmpty
+  | FieldError.TutorAboutTooLong
```
Comment (neuodev) [inline packages/sol/src/verification.ts:82]:
```suggestion
  | FieldError.InvalidTutorAbout
  | FieldError.EmptyTutorAbout
  | FieldError.TutorAboutTooLong
```

File: packages/sol/src/verification.ts [hunk 12]
```diff
@@ -0,0 +1,389 @@
...
+  | true {
+  const tutorAboutText = getTextFromHTMLString(tutorAbout);
+
+  if (!tutorAboutText.length) return FieldError.TutorAboutEmpty;
+  if (!REGEXFORHTML.test(tutorAbout)) return FieldError.TutorAboutInvalidHTML;
+  if (tutorAboutText.length > 1000) return FieldError.TutorAboutTooLong;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:88]: Always remember not to use magic numbers

Comment (neuodev) [inline packages/sol/src/verification.ts:88]:
```suggestion
  if (!tutorAboutText.length) return FieldError.EmptyToutorAbout;
  if (!REGEXFORHTML.test(tutorAbout)) return FieldError.InvalidTutorAbout;
  if (tutorAboutText.length > 1000) return FieldError.TutorAboutTooLong;
```

File: packages/sol/src/verification.ts [hunk 13]
```diff
@@ -0,0 +1,389 @@
...
+  return true;
+}
+
+export function isValidTutorNotice(
+  tutorNotice: number
+): FieldError.TutorNoticeTooLong | FieldError.TutorNoticeMinus | true {
+  if (tutorNotice < 0) return FieldError.TutorNoticeMinus;
+  if (tutorNotice > 24 * 60) return FieldError.TutorNoticeTooLong;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:96]: Always remember not to use magic numbers.

Comment (neuodev) [inline packages/sol/src/verification.ts:96]: Please think thoughtfully about your naming. Don't just name things and join words. Make sure it makes sense. 
```suggestion
  if (tutorNotice < 0) return FieldError.InvalidNotice;
  if (tutorNotice > 24 * 60) return FieldError.MaxNoticeExeceded;
```

File: packages/sol/src/verification.ts [hunk 14]
```diff
@@ -0,0 +1,389 @@
...
+  return true;
+}
+
+export function isValidRuleTitle(
+  ruleTitle: string
```
Comment (neuodev) [inline packages/sol/src/verification.ts:101]: Short names are preferred. Try to be something in the middle. Don't be very verbose and very vague. 
 
```suggestion
  title: string
```

File: packages/sol/src/verification.ts [hunk 15]
```diff
@@ -0,0 +1,389 @@
...
+):
+  | FieldError.RuleTitleInvlaidHTML
+  | FieldError.RuleTitleTooShort
+  | FieldError.RuleTitleTooLong
```
Comment (neuodev) [inline packages/sol/src/verification.ts:105]: The rule title is a plain text. It is not HTML.

File: packages/sol/src/verification.ts [hunk 16]
```diff
@@ -0,0 +1,389 @@
...
+  | true {
+  const ruleTitleText = getTextFromHTMLString(ruleTitle);
+
+  if (!REGEXFORHTML.test(ruleTitle)) return FieldError.RuleTitleInvlaidHTML;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:109]: This is not needed. Rule title is a plain string.

File: packages/sol/src/verification.ts [hunk 17]
```diff
@@ -0,0 +1,389 @@
...
+  if (ruleTitleText.length > 255) return FieldError.RuleTitleTooLong;
+  if (ruleTitleText.length < 5) return FieldError.RuleTitleTooShort;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:111]: Please don't use magic numbers.

**Review by neuodev — CHANGES_REQUESTED** [review]

File: packages/sol/src/verification.ts [hunk 18]
```diff
@@ -0,0 +1,408 @@
...
+  if (password.length > MAX_PASSWORD_LENGTH) return FieldError.PasswordTooShort;
+  if (password.length < MIN_PASSWORD_LENGTH) return FieldError.PasswordTooShort;
+  if (PASSWORD_REGEX.test(password)) return true;
+  return true;
+}
+
+export function isValidUserName(
+  userName: string
+): FieldError.InvalidUserName | true {
+  if (!USER_NAME_REGEX.test(userName)) return FieldError.InvalidUserName;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:74]: Add more analysis to the username. 
- Incase the username includes numbers or special letters return `FieldError.UserNameSpecialLettersNotAllowed` and `FieldError.UserNameNumbersNotAllowed`
- Incase the username includes English characters return `FieldError.UserNameEnglishLettersNotAllowed`

Returning these errors will help guild the user to the correct answer.

File: packages/sol/src/verification.ts [hunk 19]
```diff
@@ -0,0 +1,408 @@
...
+  return true;
+}
+
+export function isValidUserBirthYear(
+  userBirthYear: number
```
Comment (neuodev) [inline packages/sol/src/verification.ts:79]:
```suggestion
  birthYear: number
```

File: packages/sol/src/verification.ts [hunk 20]
```diff
@@ -0,0 +1,408 @@
...
+  if (userBirthYear < minYear) return FieldError.UserTooOld;
+  if (userBirthYear > maxYear) return FieldError.UserTooYoung;
+  return true;
+}
+
+export function isValidTutorBio(
+  bio: string
+): FieldError.EmptyBio | FieldError.TooLongBio | true {
+  if (!bio.length) return FieldError.EmptyBio;
+  if (bio.length > MAX_TUTOR_BIO_TEXT_LENGTH) return FieldError.TooLongBio;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:94]: Add extra check to ensure that the bio is an "Arabic" text. It should not include numbers or special letters.

File: packages/sol/src/verification.ts [hunk 21]
```diff
@@ -0,0 +1,408 @@
...
+  about: string
+):
+  | FieldError.InvalidTutorAbout
+  | FieldError.EmptyTutorAbout
+  | FieldError.TooLongTutorAbout
+  | true {
+  const tutorAboutText = getSafeInnerHtmlText(about);
+
+  if (!tutorAboutText.length) return FieldError.EmptyTutorAbout;
+  if (!HTML_REGEX.test(about)) return FieldError.InvalidTutorAbout;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:109]: The final result of `getSafeInnerHtmlText` is a plain text. This mean that this check is not necessary.

File: packages/sol/src/verification.ts [hunk 22]
```diff
@@ -0,0 +1,408 @@
...
+  return true;
+}
+
+export function isValidPassword(
+  password: string
+):
+  | FieldError.PasswordTooShort
+  | FieldError.PasswordTooLong
+  | FieldError.PasswordMissingLetters
+  | FieldError.PasswordMissingNumbers
```
Comment (neuodev) [inline packages/sol/src/verification.ts:59]:
```suggestion
  | FieldError.ShortPassword
  | FieldError.LongPassword
  | FieldError.MissingPasswordLetters
  | FieldError.MissingPasswordNumbers
```

File: packages/sol/src/verification.ts [hunk 23]
```diff
@@ -0,0 +1,408 @@
...
+
+export function isValidUserBirthYear(
+  userBirthYear: number
+): FieldError.UserTooOld | FieldError.UserTooYoung | true {
+  const currentYear = new Date().getFullYear();
+  const minYear = currentYear - MAX_USER_AGE;
+  const maxYear = currentYear - MIN_USER_AGE;
+
+  if (userBirthYear < minYear) return FieldError.UserTooOld;
+  if (userBirthYear > maxYear) return FieldError.UserTooYoung;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:86]:
```suggestion
  if (userBirthYear < minYear) return FieldError.OldUser;
  if (userBirthYear > maxYear) return FieldError.YoungUser;
```

File: packages/sol/src/verification.ts [hunk 24]
```diff
@@ -0,0 +1,408 @@
...
+  return true;
+}
+
+export function isValidTutorBio(
+  bio: string
+): FieldError.EmptyBio | FieldError.TooLongBio | true {
```
Comment (neuodev) [inline packages/sol/src/verification.ts:92]:
```suggestion
): FieldError.EmptyBio | FieldError.LongBio | true {
```

File: packages/sol/src/verification.ts [hunk 25]
```diff
@@ -0,0 +1,408 @@
...
+
+  return true;
+}
+
+export function isValidTutorAbout(
+  about: string
+):
+  | FieldError.InvalidTutorAbout
+  | FieldError.EmptyTutorAbout
+  | FieldError.TooLongTutorAbout
```
Comment (neuodev) [inline packages/sol/src/verification.ts:104]:
```suggestion
  | FieldError.LongTutorAbout
```

File: packages/sol/src/verification.ts [hunk 26]
```diff
@@ -0,0 +1,408 @@
...
+}
+
+export function isValidTutorNotice(
+  tutorNotice: number
+): FieldError.MaxNoticeExceeded | FieldError.InvalidNotice | true {
+  if (tutorNotice <= MIN_TUTOR_NOTICE_DURATION) return FieldError.InvalidNotice;
+  if (tutorNotice > MAX_TUTOR_NOTICE_DURATION)
+    return FieldError.MaxNoticeExceeded;
+  return true;
+}
```
Comment (neuodev) [inline packages/sol/src/verification.ts:122]: Couple of improvements
- Add JSDocs to indicate that the input param should be in minutes and not other units. 
- User shorter variable names. 

```suggestion
/**
* Validate tutor notice period. 
* @param {notice} input notice period in minutes. 
*/
export function isValidTutorNotice(
  notice: number
): FieldError.MaxNoticeExceeded | FieldError.InvalidNotice | true {
  if (notice <= MIN_TUTOR_NOTICE_DURATION) return FieldError.InvalidNotice;
  if (notice > MAX_TUTOR_NOTICE_DURATION)
    return FieldError.MaxNoticeExceeded;
  return true;
}
```

File: packages/sol/src/verification.ts [hunk 27]
```diff
@@ -0,0 +1,408 @@
...
+
+export function isValidRuleTitle(
+  title: string
```
Comment (neuodev) [inline packages/sol/src/verification.ts:125]: Add extra check to validate the following: 
- Title must be Arabic text.
- Can include numbers and special characters. 
- Cannot include English characters.

File: packages/sol/src/verification.ts [hunk 28]
```diff
@@ -0,0 +1,408 @@
...
+): FieldError.TooShortRuleTitle | FieldError.TooLongRuleTitle | true {
+  if (title.length > MAX_RULE_TITLE_LENGTH) return FieldError.TooLongRuleTitle;
+  if (title.length < MIN_RULE_TITLE_LENGTH) return FieldError.TooShortRuleTitle;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:128]:
```suggestion
  if (title.length > MAX_RULE_TITLE_LENGTH) return FieldError.LongRuleTitle;
  if (title.length < MIN_RULE_TITLE_LENGTH) return FieldError.ShortRuleTitle;
```

File: packages/sol/src/verification.ts [hunk 29]
```diff
@@ -0,0 +1,408 @@
...
+  | FieldError.RuleStartAfterEnd
+  | FieldError.RuleStartDatePassed
+  | true {
+  if (!Date.parse(ruleStart)) return FieldError.ISOInvalidRuleStart;
+  if (!Date.parse(ruleEnd)) return FieldError.ISOInvalidRuleEnd;
+  if (new Date(ruleStart) > new Date(ruleEnd))
+    return FieldError.RuleStartAfterEnd;
+  if (new Date() > new Date(ruleStart)) return FieldError.RuleStartDatePassed;
+  return true;
+}
```
Comment (neuodev) [inline packages/sol/src/verification.ts:147]:
```suggestion
/**
* Validate rule start date 
* @param {start} rule start date in ISO date format. 
* @param {end} rule end date in ISO date format.
*/
export function isValidRuleStart(
  start: string,
  end: string
):
  | FieldError.InvalidRuleStart
  | FieldError.InvalidRuleEnd
  | FieldError.RuleStartAfterEnd
  | FieldError.RuleStartDatePassed
  | true {
  const invalidStart = Number.isNaN(start)
  const invalidEnd = Number.isNaN(end)
  if (invalidStart) return FieldError.InvalidRuleStartFormat;
  if (invalidEnd) return FieldError.InvalidRuleEndFormat;
  if (new Date(start) >= new Date(end))
    return FieldError.RuleStartAfterEnd;
  if (new Date() > new Date(ruleStart)) return FieldError.RuleStartDatePassed;
  return true;
}
```

File: packages/sol/src/verification.ts [hunk 30]
```diff
@@ -0,0 +1,408 @@
...
+  ruleEnd: string
+):
+  | FieldError.ISOInvalidRuleStart
+  | FieldError.ISOInvalidRuleEnd
+  | FieldError.RuleStartAfterEnd
+  | FieldError.RuleStartDatePassed
+  | true {
+  if (!Date.parse(ruleStart)) return FieldError.ISOInvalidRuleStart;
+  if (!Date.parse(ruleEnd)) return FieldError.ISOInvalidRuleEnd;
+  if (new Date(ruleStart) > new Date(ruleEnd))
```
Comment (neuodev) [inline packages/sol/src/verification.ts:143]: Note that rule start and end should be plain dates and should not include time (hours, minutes, seconds). 

For example, consider these to inputs to your function `isValidRuleStart("2024-11-3 10:00", "2024-11-3 11:00")`

Your function will indicate that these are valid rule times but in reality they are the same **date**. You should only compare the dates not the datetime.  

Add regex to validate the input date format `YYYY-MM-DD`.

File: packages/sol/src/verification.ts [hunk 31]
```diff
@@ -0,0 +1,408 @@
...
+    return FieldError.RuleStartAfterEnd;
+  if (new Date() > new Date(ruleStart)) return FieldError.RuleStartDatePassed;
+  return true;
+}
+
+export function isValidRuleEnd(
```
Comment (neuodev) [inline packages/sol/src/verification.ts:149]: what we said about `isValidRuleEnd` still applies here as well.

File: packages/sol/src/verification.ts [hunk 32]
```diff
@@ -0,0 +1,408 @@
...
+  const ruleEndTime = new Date(ruleEnd).getTime();
+  const ruleStartTime = new Date(ruleStart).getTime();
+
+  if (!parsedRuleEnd) return FieldError.ISOInvalidRuleEnd;
+  if (!parsedRuleStart) return FieldError.ISOInvalidRuleStart;
+  if (ruleEndTime < ruleStartTime) return FieldError.RuleEndBeforeStart;
+  if (ruleEndTime < currentTime) return FieldError.RuleEndDatePassed;
+  return true;
+}
+
```
Comment (neuodev) [inline packages/sol/src/verification.ts:170]: Add a description to the function to indicate that the duration param should be in minutes and not other time units.

File: packages/sol/src/constants.ts [hunk 33]
```diff
@@ -1,3 +1,47 @@
 export const MINUTES_IN_HOUR = 60;
+export const MINUTES_IN_DAY = 60 * 24;
+export const MINUTES_IN_WEEK = 60 * 24 * 7;
 export const MIN_AGE = 10;
 export const MAX_AGE = 70;
+export const MAX_RULE_DURATION_MINUTES = 8 * 60;
+export const MIN_RULE_DURATION_MINUTES = 0;
```
Comment (neuodev) [inline packages/sol/src/constants.ts:7]: The minimum duration is 15 minutes which is enough to make on short lesson of 15 minutes.

File: packages/sol/src/verification.ts [hunk 34]
```diff
@@ -0,0 +1,408 @@
...
+  if (feedbackText.length > MAX_FEEDBACK_TEXT_LENGTH)
+    return FieldError.TooLongInterviewFeedback;
+  return true;
+}
+export function isValidInterviewNote(
+  note: string
+):
+  | FieldError.TooLongInterviewNote
+  | FieldError.TooShortInterviewNote
+  | FieldError.InvalidInterviewNote
```
Comment (neuodev) [inline packages/sol/src/verification.ts:202]:
```suggestion
  | FieldError.LongInterviewNote
  | FieldError.ShortInterviewNote
  | FieldError.InvalidInterviewNote
```

File: packages/sol/src/verification.ts [hunk 35]
```diff
@@ -0,0 +1,408 @@
...
+  | true {
+  const noteText = note.replace(HTML_TAGS_REGEX, "");
```
Comment (neuodev) [inline packages/sol/src/verification.ts:204]: Use `getSafeInnerHtml` here as well. 
```suggestion
const text = getSafeInnerHtml(note);
```

File: packages/sol/src/verification.ts [hunk 36]
```diff
@@ -0,0 +1,408 @@
...
+  if (noteText.length > MAX_NOTE_TEXT_LENGTH)
+    return FieldError.TooLongInterviewNote;
+  return true;
+}
+
+export function isValidInterviewLevel(
+  level: number
+): FieldError.TooLowInterviewLevel | FieldError.TooHighInterviewLevel | true {
+  if (level < MIN_INTERVIEW_LEVEL) return FieldError.TooLowInterviewLevel;
+  if (level > MAX_INTERVIEW_LEVEL) return FieldError.TooHighInterviewLevel;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:218]:
```suggestion
  if (level < MIN_INTERVIEW_LEVEL) return FieldError.LowInterviewLevel;
  if (level > MAX_INTERVIEW_LEVEL) return FieldError.HighInterviewLevel;
```

File: packages/sol/src/verification.ts [hunk 37]
```diff
@@ -0,0 +1,408 @@
...
+  return true;
+}
+
+export function isValidRatingValue(
+  ratingValue: number
+): FieldError.TooLowRatingValue | FieldError.TooHighRatingValue | true {
+  if (ratingValue < MIN_RATING_VALUE) return FieldError.TooLowRatingValue;
+  if (ratingValue > MAX_RATING_VALUE) return FieldError.TooHighRatingValue;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:226]:
```suggestion
  if (ratingValue < MIN_RATING_VALUE) return FieldError.LowRatingValue;
  if (ratingValue > MAX_RATING_VALUE) return FieldError.HighRatingValue;
```

File: packages/sol/src/verification.ts [hunk 38]
```diff
@@ -0,0 +1,408 @@
...
+
+export function isValidRatingText(
+  ratingText: string
+): FieldError.TooLongRatingText | FieldError.TooShortRatingText | true {
+  if (ratingText.length > MAX_RATING_TEXT_LENGTH)
+    return FieldError.TooLongRatingText;
+  if (ratingText.length < MIN_RATING_TEXT_LENGTH)
+    return FieldError.TooShortRatingText;
+  return true;
+}
```
Comment (neuodev) [inline packages/sol/src/verification.ts:238]:
```suggestion
export function isValidRatingText(
  rating: string
): FieldError.LongRatingText | FieldError.ShortRatingText | true {
  if (rating.length > MAX_RATING_TEXT_LENGTH)
    return FieldError.TooLongRatingText;
  if (rating.length <= MIN_RATING_TEXT_LENGTH)
    return FieldError.ShortRatingText;
  return true;
}
```

File: packages/sol/src/verification.ts [hunk 39]
```diff
@@ -0,0 +1,408 @@
...
+
+export function isValidPlanAlias(
+  planAlias: string
+): FieldError.TooLongPlanAlias | FieldError.TooShortPlanAlias | true {
```
Comment (neuodev) [inline packages/sol/src/verification.ts:242]: The plan alias should be a plain Arabic text. Numbers and especial characters are allowed.  

```suggestion
export function isValidPlanAlias(
  alias: string
): FieldError.LongPlanAlias | FieldError.ShortPlanAlias | true {
```

File: packages/sol/src/verification.ts [hunk 40]
```diff
@@ -0,0 +1,408 @@
...
+  if (planAlias.length > MAX_PLAN_ALIAS_LENGTH)
+    return FieldError.TooLongPlanAlias;
+  if (planAlias.length < MIN_PLAN_ALIAS_LENGTH)
+    return FieldError.TooShortPlanAlias;
+  return true;
+}
```
Comment (neuodev) [inline packages/sol/src/verification.ts:248]: Add empty line between functions. 
```suggestion
}

```

File: packages/sol/src/verification.ts [hunk 41]
```diff
@@ -0,0 +1,408 @@
...
+export function isValidPlanWeeklyMinutes(
```
Comment (neuodev) [inline packages/sol/src/verification.ts:249]: Add description to indicate the input to the function should be in minutes and not other unites.

File: packages/sol/src/verification.ts [hunk 42]
```diff
@@ -0,0 +1,408 @@
...
+export function isValidMessageText(
+  message: string
+):
+  | FieldError.InvalidMessageText
+  | FieldError.TooLongMessageText
+  | FieldError.TooShortMessageText
+  | true {
+  const text = getSafeInnerHtmlText(message);
+
+  if (!HTML_REGEX.test(message)) return FieldError.InvalidMessageText;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:327]: This check can be safely removed.

File: packages/sol/src/verification.ts [hunk 43]
```diff
@@ -0,0 +1,408 @@
...
+  if (text.length > MAX_MESSAGE_TEXT_LENGTH)
+    return FieldError.TooLongMessageText;
+  if (text.length < MIN_MESSAGE_TEXT_LENGTH)
+    return FieldError.TooShortMessageText;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:331]:
```suggestion
    return FieldError.LongMessageText;
  if (text.length < MIN_MESSAGE_TEXT_LENGTH)
    return FieldError.ShortMessageText;
```

**Review by neuodev — CHANGES_REQUESTED** [review]

File: packages/sol/src/verification.ts (hunk 27 above)
Comment (moalidv) [reply packages/sol/src/verification.ts:125]: @neuodev  
if we must write it in arabic

File: packages/sol/src/verification.ts [hunk 44]
```diff
@@ -0,0 +1,487 @@
...
+export function isValidInterviewFeedback(
+  feedback: string
+):
+  | FieldError.TooLongInterviewFeedback
+  | FieldError.TooShortInterviewFeedback
+  | FieldError.InvalidInterviewFeedback
+  | true {
+  const feedbackText = getSafeInnerHtmlText(feedback);
+
+  if (!HTML_REGEX.test(feedback)) return FieldError.InvalidInterviewFeedback;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:271]: Please remove it.

File: packages/sol/src/verification.ts [hunk 45]
```diff
@@ -0,0 +1,487 @@
...
+export function isValidInterviewNote(
+  note: string
+):
+  | FieldError.LongInterviewNote
+  | FieldError.ShortInterviewNote
+  | FieldError.InvalidInterviewNote
+  | true {
+  const text = getSafeInnerHtmlText(note);
+
+  if (!HTML_REGEX.test(note)) return FieldError.InvalidInterviewNote;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:287]: Please remove

File: packages/sol/src/verification.ts [hunk 46]
```diff
@@ -0,0 +1,487 @@
...
+  rating: string
+): FieldError.LongRatingText | FieldError.ShortRatingText | true {
+  if (rating.length > MAX_RATING_TEXT_LENGTH) return FieldError.LongRatingText;
+  if (rating.length <= MIN_RATING_TEXT_LENGTH)
+    return FieldError.ShortRatingText;
+  return true;
+}
+
+export function isValidPlanAlias(
+  alias: string
```
Comment (neuodev) [inline packages/sol/src/verification.ts:319]: Alias should: 
1. Arabic 
2. Special Characters. 
3. Numbers.

File: packages/sol/src/verification.ts [hunk 47]
```diff
@@ -0,0 +1,487 @@
...
+  discount: number
+):
+  | FieldError.ZeroPlanPrice
+  | FieldError.InfinitePlanPrice
+  | FieldError.PlanTotalDiscount
+  | FieldError.PlanPriceNotInteger
+  | true {
+  const priceAfterDiscount = price - (price * discount) / 100;
+
+  if (price <= 0) return FieldError.ZeroPlanPrice;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:354]:
```suggestion
  if (price <= 0) return FieldError.ZeroOrNegativePlanPrice;
```

File: packages/sol/src/verification.ts [hunk 48]
```diff
@@ -0,0 +1,487 @@
...
+  if (!Number.isFinite(price)) return FieldError.InfinitePlanPrice;
+  if (priceAfterDiscount <= 0) return FieldError.PlanTotalDiscount;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:356]:
```suggestion
  if (priceAfterDiscount < 0) return FieldError.NegaitvePriceAfterDiscount;
```

File: packages/sol/src/verification.ts [hunk 49]
```diff
@@ -0,0 +1,487 @@
...
+}
+export function isValidPlanDiscount(
+  discount: number
+):
+  | FieldError.MinPlanDiscountSubceeded
+  | FieldError.PlanTotalDiscount
+  | FieldError.MaxPlanDiscountExceeded
+  | true {
+  if (discount < MIN_PLAN_DISCOUNT) return FieldError.MinPlanDiscountSubceeded;
+  if (discount == MIN_PLAN_DISCOUNT) return FieldError.PlanTotalDiscount;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:368]:
```suggestion
```

File: packages/sol/src/verification.ts [hunk 50]
```diff
@@ -0,0 +1,487 @@
...
+  if (discount > MAX_PLAN_DISCOUNT) return FieldError.MaxPlanDiscountExceeded;
+  return true;
+}
+
+export function isValidCouponCode(
+  couponCode: string
```
Comment (neuodev) [inline packages/sol/src/verification.ts:374]:
```suggestion
  code: string
```

File: packages/sol/src/verification.ts [hunk 51]
```diff
@@ -0,0 +1,487 @@
...
+): FieldError.InvalidCouponCode | true {
+  if (!COUPON_REGEX.test(couponCode)) return FieldError.InvalidCouponCode;
+  return true;
+}
+export function isValidCouponDiscount(
+  couponDiscount: number
```
Comment (neuodev) [inline packages/sol/src/verification.ts:380]:
```suggestion
  discount: number
```

File: packages/sol/src/verification.ts [hunk 52]
```diff
@@ -0,0 +1,487 @@
...
+): FieldError.ZeroCouponDiscount | FieldError.MaxCouponDiscountExceeded | true {
+  if (couponDiscount <= MIN_DISCOUNT_VALUE)
+    return FieldError.ZeroCouponDiscount;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:383]:
```suggestion
  if (discount < MIN_DISCOUNT_VALUE)
    return FieldError.NegativeDiscount;
```

File: packages/sol/src/verification.ts [hunk 53]
```diff
@@ -0,0 +1,487 @@
...
+}
+export function isValidCouponExpireDate(
+  expireDate: string
+):
+  | FieldError.ExpiredCouponExpireDate
+  | FieldError.InvalidCouponExpireDate
+  | true {
+  if (!Date.parse(expireDate)) return FieldError.InvalidCouponExpireDate;
+  if (new Date() > new Date(expireDate))
+    return FieldError.ExpiredCouponExpireDate;
```
Comment (neuodev) [inline packages/sol/src/verification.ts:396]:
```suggestion
    return FieldError.CouponExpired;
```

**Review by neuodev — CHANGES_REQUESTED** [review]

**Comment by neuodev** [comment]
🚀 Nova: https://nova-o1yok9d29-litespace-cores-projects.vercel.app
🚀 Luna: https://luna-2c3mgbslc-litespace-cores-projects.vercel.app
🚀 Nebula: https://nebula-kyxuff2i3-litespace-cores-projects.vercel.app
🚀 Dashboard: https://dashboard-3hvhxfhfh-litespace-cores-projects.vercel.app

**Pull request merged**